File: fetch.py
#!/usr/bin/env python3
"""
freetvco – IPTV stream fetcher for Cò TiVi
============================================
Reverse-engineered from APK: com.cofvuong.tivi v1.1.7 (Hermes bytecode)

Encryption (confirmed via hermes-dec disassembly of Fn#11130/11461 'giaimane'):
  Algorithm : AES-256-ECB
  Padding   : PKCS7
  Key/Channels : 6677150266771502cotivichatvkl321  (32 bytes)
  Key/Sports   : 6677150266771502cotivichatvkl999  (32 bytes)

API:
  GET https://api.cotivi.site/api/Channels?version=1.1.2
  GET https://api.cotivi.site/api/Sports?version=1.1.2
  Response: { "key": "<base64 troll>", "data": "<base64 AES-ECB ciphertext>" }
"""

import base64, json, os, sys, urllib.request
from pathlib import Path
from datetime import datetime, timezone, timedelta

# ── pycryptodome ──────────────────────────────────────────────────────────────
try:
    from Crypto.Cipher import AES
    from Crypto.Util.Padding import unpad
except ImportError:
    sys.exit("❌  pip install pycryptodome")

# ── config ────────────────────────────────────────────────────────────────────
API_BASE = "https://api.cotivi.site"
VERSION  = os.environ.get("COTIVI_VERSION", "1.1.2")
OUT_DIR  = Path(os.environ.get("OUT_DIR",   "output"))
DBG_DIR  = Path(os.environ.get("DEBUG_DIR", "debug"))

# Keys extracted from Hermes bytecode (fn#11130 'giaimane', fn#11138 '?anon_0_')
ENDPOINTS = [
    {
        "name"    : "channels",
        "path"    : "/api/Channels",
        "key"     : b"6677150266771502cotivichatvkl321",   # 32 bytes → AES-256
        "outFile" : "cotivi_channels.m3u",
        "label"   : "📺 Channels",
    },
    {
        "name"    : "sports",
        "path"    : "/api/Sports",
        "key"     : b"6677150266771502cotivichatvkl999",   # 32 bytes → AES-256
        "outFile" : "cotivi_sports.m3u",
        "label"   : "⚽ Sports",
    },
]

# ── helpers ───────────────────────────────────────────────────────────────────

def fetch_json(url: str) -> dict:
    req = urllib.request.Request(url, headers={"User-Agent": "Cò TiVi/1.1.7"})
    with urllib.request.urlopen(req, timeout=30) as r:
        return json.loads(r.read())


def decrypt_aes_ecb(b64_data: str, key: bytes) -> dict:
    """AES-256-ECB + PKCS7 — matches CryptoJS call in the app's giaimane()."""
    raw    = base64.b64decode(b64_data)
    cipher = AES.new(key, AES.MODE_ECB)
    plain  = unpad(cipher.decrypt(raw), AES.block_size)
    return json.loads(plain.decode("utf-8"))


def make_m3u(data_json: dict, group_prefix: str = "") -> str:
    lines = ["#EXTM3U"]
    for group in data_json.get("Data", []):
        grp = group.get("Kenh", "Unknown")
        for ch in group.get("List", []):
            link = ch.get("link", "")
            if not link or not link.startswith("http"):
                # Sports entries without a direct link (need separate token fetch)
                continue
            name = ch.get("name") or ch.get("id", "Unknown")
            icon = ch.get("icon", "")
            hdr  = ch.get("header", {})
            ua   = hdr.get("User-agent") or hdr.get("user-agent", "")
            lines.append(
                f'#EXTINF:-1 tvg-id="{ch.get("id","")}" '
                f'tvg-name="{name}" tvg-logo="{icon}" '
                f'group-title="{grp}",{name}'
            )
            if ua:
                lines.append(f"#EXTVLCOPT:http-user-agent={ua}")
            lines.append(link)
    return "\n".join(lines) + "\n"


# ── main ──────────────────────────────────────────────────────────────────────

def main():
    OUT_DIR.mkdir(parents=True, exist_ok=True)
    DBG_DIR.mkdir(parents=True, exist_ok=True)

    vn_now = datetime.now(timezone(timedelta(hours=7))).strftime("%Y-%m-%d %H:%M %Z")
    all_lines = ["#EXTM3U"]
    all_count = 0

    for ep in ENDPOINTS:
        url = f"{API_BASE}{ep['path']}?version={VERSION}"
        print(f"\n{ep['label']} → {url}")

        try:
            resp = fetch_json(url)
        except Exception as e:
            print(f"  ❌ Fetch failed: {e}")
            continue

        # Save raw encrypted response for debugging
        raw_path = DBG_DIR / f"{ep['name']}_raw.json"
        raw_path.write_text(json.dumps(resp, ensure_ascii=False, indent=2), encoding="utf-8")

        key_field = base64.b64decode(resp.get("key", "")).decode("utf-8", errors="replace")
        print(f"  key_field (troll) : {key_field!r}")
        print(f"  data size         : {len(base64.b64decode(resp['data']))} bytes")

        try:
            plain = decrypt_aes_ecb(resp["data"], ep["key"])
        except Exception as e:
            print(f"  ❌ Decrypt failed: {e}")
            continue

        # Save decrypted JSON
        dec_path = DBG_DIR / f"{ep['name']}_decrypted.json"
        dec_path.write_text(json.dumps(plain, ensure_ascii=False, indent=2), encoding="utf-8")

        m3u_content = make_m3u(plain)
        count = m3u_content.count("#EXTINF")
        print(f"  ✅ Decrypted — {count} channels")

        # Write per-endpoint M3U
        out_path = OUT_DIR / ep["outFile"]
        out_path.write_text(m3u_content, encoding="utf-8")
        print(f"  📄 {out_path}")

        # Append to combined (skip header line)
        for line in m3u_content.split("\n")[1:]:
            if line.strip():
                all_lines.append(line)
        all_count += count

    # Write combined M3U
    if all_count > 0:
        all_m3u = "\n".join(all_lines) + "\n"
        all_path = OUT_DIR / "cotivi_all.m3u"
        all_path.write_text(all_m3u, encoding="utf-8")
        print(f"\n✅ Combined: {all_count} channels → {all_path}")
    else:
        print("\n❌ No channels decrypted")
        sys.exit(1)


if __name__ == "__main__":
    main()

File: fetch.js
#!/usr/bin/env node
/**
 * freetvco - IPTV stream fetcher for Cò TiVi
 * Fetches encrypted channel/sport data from api.cotivi.site,
 * decrypts with BlowFish (CryptoJS), and generates M3U playlist.
 *
 * App analysis:
 *  - Bundle: Hermes bytecode (React Native / Expo SDK 52)
 *  - Encryption: CryptoJS.Blowfish, padding AnsiX923
 *  - API: https://api.cotivi.site/api/Channels?version=<runtimeVersion>
 *  - Response: { key: "<base64>", data: "<base64-encrypted>" }
 *  - key field decoded: used as BlowFish key
 */

const https = require('https');
const fs   = require('fs');
const path = require('path');
const CryptoJS = require('crypto-js');

const API_BASE   = 'https://api.cotivi.site';
const VERSION    = process.env.COTIVI_VERSION || '1.1.2';
const OUT_DIR    = process.env.OUT_DIR || path.join(__dirname, '../../output');
const DEBUG_DIR  = process.env.DEBUG_DIR || path.join(__dirname, '../../debug');

// ─── helpers ─────────────────────────────────────────────────────────────────

function fetchJson(url) {
  return new Promise((resolve, reject) => {
    https.get(url, { headers: { 'User-Agent': 'Cò TiVi/1.1.7' } }, res => {
      const chunks = [];
      res.on('data', c => chunks.push(c));
      res.on('end', () => {
        try { resolve(JSON.parse(Buffer.concat(chunks).toString())); }
        catch (e) { reject(new Error(`JSON parse error from ${url}: ${e.message}`)); }
      });
    }).on('error', reject);
  });
}

/**
 * Try every combination of mode / padding / IV until we find valid JSON.
 * Returns the parsed object or null if no combination works.
 */
function tryDecrypt(keyStr, encDataB64) {
  const encDataWA = CryptoJS.enc.Base64.parse(encDataB64);

  const modes = [
    { name: 'CBC', mode: CryptoJS.mode.CBC },
    { name: 'ECB', mode: CryptoJS.mode.ECB },
    { name: 'CFB', mode: CryptoJS.mode.CFB },
  ];
  const paddings = [
    { name: 'AnsiX923',    pad: CryptoJS.pad.AnsiX923 },
    { name: 'Pkcs7',       pad: CryptoJS.pad.Pkcs7 },
    { name: 'ZeroPadding', pad: CryptoJS.pad.ZeroPadding },
    { name: 'NoPadding',   pad: CryptoJS.pad.NoPadding },
  ];
  const ivHexes = [
    '0000000000000000',
    Buffer.from(keyStr.slice(0, 8)).toString('hex'),
    '3132333435363738', // '12345678'
    Buffer.from('cotivi00').toString('hex'),
  ];

  // Key variants (WordArray and passphrase string)
  const keyWordArray = CryptoJS.enc.Utf8.parse(keyStr);
  const keyLower     = CryptoJS.enc.Utf8.parse(keyStr.toLowerCase());

  for (const { name: mName, mode } of modes) {
    for (const { name: pName, pad: padding } of paddings) {
      const ivList = mName === 'ECB' ? [null] : ivHexes;
      for (const ivHex of ivList) {
        for (const keyWA of [keyWordArray, keyLower]) {
          try {
            const opts = { mode, padding };
            if (ivHex) opts.iv = CryptoJS.enc.Hex.parse(ivHex);
            const dec = CryptoJS.Blowfish.decrypt({ ciphertext: encDataWA }, keyWA, opts);
            const text = dec.toString(CryptoJS.enc.Utf8);
            if (text && text.length > 20 && (text[0] === '{' || text[0] === '[')) {
              console.log(`✅ Decrypt OK: mode=${mName} pad=${pName} iv=${ivHex||'none'}`);
              return JSON.parse(text);
            }
          } catch (_) { /* keep trying */ }
        }
      }
    }
  }
  return null;
}

// ─── M3U builder ─────────────────────────────────────────────────────────────

/**
 * Detect stream URL field name from first channel object.
 * Common field names: url, streamUrl, stream_url, link, playUrl, liveUrl
 */
function pickUrl(ch) {
  const candidates = ['url', 'streamUrl', 'stream_url', 'link', 'playUrl', 'liveUrl',
                       'stream', 'hls', 'rtmp', 'src', 'source'];
  for (const c of candidates) if (ch[c]) return ch[c];
  // Fallback: any field containing http
  return Object.values(ch).find(v => typeof v === 'string' && v.startsWith('http')) || '';
}

function buildM3U(channels, groupTitle = 'CoTiVi') {
  let m3u = `#EXTM3U url-tvg="https://lichphatsong.site/schedule/epg.xml.gz"\n`;
  for (const ch of channels) {
    const name  = ch.name || ch.title || ch.channelName || 'Unknown';
    const logo  = ch.logo || ch.icon || ch.thumbnail || ch.image || '';
    const group = ch.group || ch.category || ch.groupName || groupTitle;
    const id    = ch.id || ch.channelId || name.toLowerCase().replace(/\s+/g, '-');
    const url   = pickUrl(ch);
    if (!url) continue;
    m3u += `#EXTINF:-1 tvg-id="${id}" tvg-name="${name}" tvg-logo="${logo}" group-title="${group}",${name}\n`;
    m3u += `${url}\n`;
  }
  return m3u;
}

// ─── main ────────────────────────────────────────────────────────────────────

async function main() {
  fs.mkdirSync(OUT_DIR,   { recursive: true });
  fs.mkdirSync(DEBUG_DIR, { recursive: true });

  const endpoints = [
    { name: 'channels', path: '/api/Channels', outFile: 'cotivi_channels.m3u', group: 'CoTiVi TV' },
    { name: 'sports',   path: '/api/Sports',   outFile: 'cotivi_sports.m3u',   group: 'CoTiVi Sports' },
  ];

  let allChannels = [];
  let decryptOk   = false;

  for (const ep of endpoints) {
    const url = `${API_BASE}${ep.path}?version=${VERSION}`;
    console.log(`\n📡 Fetching ${ep.name}: ${url}`);

    let resp;
    try { resp = await fetchJson(url); }
    catch (e) { console.error(`  ❌ Fetch error: ${e.message}`); continue; }

    // Save raw response for debugging
    fs.writeFileSync(
      path.join(DEBUG_DIR, `${ep.name}_raw.json`),
      JSON.stringify(resp, null, 2)
    );

    if (!resp.key || !resp.data) {
      console.error(`  ❌ Unexpected response format`);
      continue;
    }

    const keyStr = Buffer.from(resp.key, 'base64').toString('utf8');
    console.log(`  🔑 Key field decoded: ${keyStr} (${keyStr.length} chars)`);
    console.log(`  📦 Encrypted data size: ${Buffer.from(resp.data, 'base64').length} bytes`);

    const data = tryDecrypt(keyStr, resp.data);
    if (!data) {
      console.error(`  ❌ All decrypt attempts failed. Raw data saved to debug/${ep.name}_raw.json`);
      console.error(`     → Next step: decompile Hermes bytecode or intercept live traffic`);
      continue;
    }

    decryptOk = true;
    const channels = Array.isArray(data) ? data : (data.channels || data.data || data.items || []);
    console.log(`  ✅ Decrypted ${channels.length} entries`);

    if (channels.length === 0) {
      console.warn(`  ⚠️  No channels found in decrypted data`);
      fs.writeFileSync(path.join(DEBUG_DIR, `${ep.name}_decrypted.json`), JSON.stringify(data, null, 2));
      continue;
    }

    const m3u = buildM3U(channels, ep.group);
    const outPath = path.join(OUT_DIR, ep.outFile);
    fs.writeFileSync(outPath, m3u, 'utf8');
    console.log(`  📺 Saved ${channels.length} channels → ${outPath}`);

    // Save decrypted for inspection
    fs.writeFileSync(
      path.join(DEBUG_DIR, `${ep.name}_decrypted.json`),
      JSON.stringify(channels.slice(0, 3), null, 2)
    );

    allChannels = allChannels.concat(channels.map(ch => ({ ...ch, _group: ep.group })));
  }

  // Combined playlist
  if (allChannels.length > 0) {
    const combined = buildM3U(allChannels, 'CoTiVi');
    fs.writeFileSync(path.join(OUT_DIR, 'cotivi_all.m3u'), combined, 'utf8');
    console.log(`\n📋 Combined playlist: ${allChannels.length} channels → output/cotivi_all.m3u`);
  }

  if (!decryptOk) {
    console.error('\n❌ Decryption failed for all endpoints.');
    console.error('   Possible causes:');
    console.error('   1. Hardcoded key in Hermes bytecode not yet extracted');
    console.error('   2. Different IV / key derivation');
    console.error('   3. Non-CryptoJS BlowFish implementation');
    console.error('\n   Recommended next step: run APK in emulator with mitmproxy to intercept decrypted API calls.');
    process.exit(1);
  }

  console.log('\n✅ Done!');
}

main().catch(e => { console.error(e); process.exit(1); });
